### PR TITLE
QAT-670: Add Allure support

### PR DESCRIPTION
**Pull request by denisqsound**

### CODE REVIEW

File: test/smoke/run.sh [hunk 1]
```diff
@@ -86,10 +86,23 @@ if [ "${SKIP_CLUSTER_CREATION:-false}" = "false" ]; then
...
+  - role: control-plane
+    extraPortMappings:
+      - containerPort: 30000
+        hostPort: 8080
+        protocol: TCP
+    extraMounts:
+      - hostPath: ${PWD}/allure_report
+        containerPath: /allure_report
+EOF
+)
```
Comment (xDmitriev) [inline test/smoke/run.sh:105]: if we move kind.yaml here, we also need to cleanup kind.yaml file

File: test/smoke/run-smoke-suite.sh [hunk 2]
```diff
@@ -92,24 +92,74 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
+spec:
+  terminationGracePeriodSeconds: 0
+  restartPolicy: Never
+  imagePullSecrets:
+    - name: "${SMOKE_IMAGE_PULL_SECRET_NAME}"
+  containers:
+  - command:
+    - sleep
+    - infinity
+    env:
```
Comment (xDmitriev) [inline test/smoke/run-smoke-suite.sh:110]: can not see where you pass `ALLURE*` vars to the container. How allurectl in contaener knows where to send the report and other settings?

File: test/smoke/run-smoke-suite.sh [hunk 3]
```diff
@@ -92,24 +92,74 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
\ No newline at end of file
+
+if [ "${ALLURE_UPLOAD_REPORT:-false}" = "true" ]; then
+  RUN_TESTS='allurectl watch -- pytest'
+else
+  RUN_TESTS="pytest"
+fi
+
+echo "Executing command: kubectl exec pytest ${EXEC_ARGS} -- ${RUN_TESTS} -n ${PYTEST_WORKERS} ${PYTEST_ARGS}" && \
+kubectl exec pytest ${EXEC_ARGS} -- ${RUN_TESTS} -n "${PYTEST_WORKERS}" "${PYTEST_ARGS}"
```
Comment (xDmitriev) [inline test/smoke/run-smoke-suite.sh:165]: no quotes here. I know that linter complain)

File: test/smoke/run-smoke-suite.sh [hunk 4]
```diff
@@ -92,24 +92,74 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
+    - name: HOSTNAME_OLD_NODE
+      value: "${HOSTNAME_OLD_NODE}"
+    image: "${SMOKE_IMAGE_NAME}:${SMOKE_IMAGE_TAG}"
+    imagePullPolicy: IfNotPresent
+    name: pytest
+EOF
+)
+
+if [ "${ALLURE_GENERATE_REPORT:-false}" = "true" ]; then
+POD_CONFIG+=$(cat << EOF
```
Comment (xDmitriev) [inline test/smoke/run-smoke-suite.sh:134]: Nice trick))

File: .env.example [hunk 5]
```diff
@@ -12,7 +12,7 @@ USER_UUID=...
 USER_SECRET=...
```
Comment (xDmitriev) [inline .env.example:12]: Can not see CI changes

File: test/smoke/run.sh (hunk 1 above)
Comment (denisqsound) [reply test/smoke/run.sh:105]: removed kind.yaml file

File: test/smoke/run-smoke-suite.sh [hunk 6]
```diff
@@ -92,24 +92,73 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
-kubectl exec pytest ${EXEC_ARGS} -- pytest -n ${PYTEST_WORKERS} ${PYTEST_ARGS}
\ No newline at end of file
+
+if [ "${ALLURE_UPLOAD_REPORT:-false}" = "true" ]; then
+  RUN_TESTS='allurectl watch -- pytest'
+else
+  RUN_TESTS="pytest"
+fi
+
+kubectl exec pytest ${EXEC_ARGS} -- "${RUN_TESTS}" -n "${PYTEST_WORKERS}" "${PYTEST_ARGS}"
```
[the remark this answers is not in the document]
Comment (denisqsound) [reply test/smoke/run-smoke-suite.sh]: fixed

File: test/smoke/run-smoke-suite.sh (hunk 3 above)
Comment (denisqsound) [reply test/smoke/run-smoke-suite.sh]: remove everywhere quotes

File: .env.example [hunk 7]
```diff
@@ -25,6 +25,18 @@ USER_SECRET=...
 SMOKE_REGISTRY_TOKEN=...
 SMOKE_REGISTRY_SECRET=...
 
+# Environment variables for Allure
```
[the remark this answers is not in the document]
Comment (denisqsound) [reply .env.example:28]: here - https://github.com/wallarm/ingress/blob/e8fb4154ca14bd770c8a0db3e04d79faf8c0e1b2/test/smoke/run-smoke-suite.sh#L159

File: test/smoke/run-smoke-suite.sh (hunk 2 above)
Comment (denisqsound) [reply test/smoke/run-smoke-suite.sh:110]: this place - https://github.com/wallarm/ingress/blob/e8fb4154ca14bd770c8a0db3e04d79faf8c0e1b2/test/smoke/run-smoke-suite.sh#L165-L173

File: .env.example (hunk 5 above)
Comment (denisqsound) [reply .env.example:12]: CI and local run works equally
if use ALLURE_GENERATE_REPORT - report will generated
ALLURE_UPLOAD_REPORT - report will upload on TestOps

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -145,9 +144,21 @@ jobs:
           USER_UUID: ${{ secrets.USER_UUID }}
           SMOKE_REGISTRY_TOKEN: ${{ secrets.SMOKE_REGISTRY_TOKEN }}
           SMOKE_REGISTRY_SECRET: ${{ secrets.SMOKE_REGISTRY_SECRET }}
+          ALLURE_GENERATE_REPORT: true
+          ALLURE_UPLOAD_REPORT: true
+          ALLURE_SERVER_TOKEN: ${{ secrets.ALLURE_SERVER_TOKEN }}
+          ALLURE_SERVER_URL: ${{ secrets.ALLURE_SERVER_URL }}
+          ALLURE_PROJECT_ID: ${{ secrets.ALLURE_PROJECT_ID }}
+          SECRET: ${{ secrets.PGP_SECRET_SIGNING_PASSPHRASE }}
+          ARCH: ${{ matrix.ARCH }}
```
[the remark this answers is not in the document]
Comment (denisqsound) [reply .github/workflows/ci.yaml]: fixed

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -145,9 +144,21 @@ jobs:
...
         run: |
           make kind-smoke-test
 
+      - name: Delete cluster
```
[the remark this answers is not in the document]
Comment (denisqsound) [reply .github/workflows/ci.yaml]: removed

File: .env.example [hunk 10]
```diff
@@ -25,6 +25,18 @@ USER_SECRET=...
...
+ALLURE_GENERATE_REPORT=...# true or false (default false)
+
+# If you want to upload report to Allure server
+ALLURE_UPLOAD_REPORT=...# true or false (default false)
+ALLURE_SERVER_URL=...# (default https://allure.wallarm.com)
```
[the remark this answers is not in the document]
Comment (denisqsound) [reply .env.example]: moved

File: test/smoke/run-smoke-suite.sh [hunk 11]
```diff
@@ -92,24 +92,108 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
+      value: "${USER_SECRET}"
+    - name: HOSTNAME_OLD_NODE
+      value: "${HOSTNAME_OLD_NODE}"
+    image: "${SMOKE_IMAGE_NAME}:${SMOKE_IMAGE_TAG}"
+    imagePullPolicy: IfNotPresent
+    name: pytest
+EOF
+)
+
+if [ "${ALLURE_GENERATE_REPORT:-false}" = "true" ]; then
```
Comment (xDmitriev) [inline test/smoke/run-smoke-suite.sh:133]: Let's remove conditional logic. It becomes more and more harder to read the code.

File: test/smoke/run-smoke-suite.sh [hunk 12]
```diff
@@ -92,24 +92,108 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
+
+ENV_VARS=""
+GITHUB_VARS=$(env | grep "^GITHUB_" | awk -F '=' '{print $1 "=${" $1 "}" }')
+if [ "${ALLURE_UPLOAD_REPORT:-false}" = "true" ]; then
+  set -x
+  env
+  export ALLURE_ENDPOINT="${ALLURE_ENDPOINT:-https://allure.wallarm.com}"
+  export ALLURE_PROJECT_ID="${ALLURE_PROJECT_ID:-24}"
+  export ALLURE_TOKEN="${ALLURE_TOKEN:-${ALLURE_SERVER_TOKEN}}"
+  export ALLURE_LAUNCH_NAME="${GITHUB_WORKFLOW:-local}-${GITHUB_RUN_ID:-local}-${GITHUB_HEAD_REF:-local}-${GITHUB_JOB:-local}"
```
Comment (xDmitriev) [inline test/smoke/run-smoke-suite.sh:165]: make lunch name the same as we agreed[ here ](https://wallarm.slack.com/archives/C055NF0CMNJ/p1697720136053809?thread_ts=1697707514.344089&cid=C055NF0CMNJ)

File: test/smoke/run-smoke-suite.sh [hunk 13]
```diff
@@ -92,24 +92,108 @@ NODE_UUID=$(kubectl exec "${POD}" -c controller -- cat /opt/wallarm/etc/wallarm/
...
+kubectl wait --for=condition=Ready pods --all --timeout=600s
 
 echo "Run smoke tests ..."
 trap get_logs_and_fail ERR
-kubectl exec pytest ${EXEC_ARGS} -- pytest -n ${PYTEST_WORKERS} ${PYTEST_ARGS}
\ No newline at end of file
+
+ENV_VARS=""
+GITHUB_VARS=$(env | grep "^GITHUB_" | awk -F '=' '{print $1 "=${" $1 "}" }')
+if [ "${ALLURE_UPLOAD_REPORT:-false}" = "true" ]; then
```
Comment (xDmitriev) [inline test/smoke/run-smoke-suite.sh:159]: Remove conditional logic as I mentioned above

File: test/smoke/run.sh [hunk 14]
```diff
@@ -86,10 +86,23 @@ if [ "${SKIP_CLUSTER_CREATION:-false}" = "false" ]; then
...
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+  - role: control-plane
+    extraPortMappings:
+      - containerPort: 30000
+        hostPort: 8080
+        protocol: TCP
+    extraMounts:
+      - hostPath: ${PWD}/allure_report
```
Comment (xDmitriev) [inline test/smoke/run.sh:102]: CURDIR instead of PWD everywhere
Remove PWD from script

File: .github/workflows/smoke-test.yaml [hunk 15]
```diff
@@ -22,7 +22,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        k8s: [ v1.23.13, v1.27.1 ]
+        k8s: [ v1.23.13, v1.24.15, v1.25.11, v1.26.6, v1.27.1 ]
```
Comment (xDmitriev) [inline .github/workflows/smoke-test.yaml:25]: keep just 3 versions here: v1.23.13, v1.25.11 and v1.27.1

File: .github/workflows/ci.yaml [hunk 16]
```diff
@@ -109,33 +106,38 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        k8s: [v1.23.13, v1.27.1]
-        ARCH: [amd64, arm64]
+        k8s: [ v1.23.13, v1.24.15, v1.25.11, v1.26.6, v1.27.1 ]
+        ARCH: [ amd64 ]
         include:
           - ARCH: amd64
             RUNNER: self-hosted-2cpu
```
Comment (xDmitriev) [inline .github/workflows/ci.yaml:113]:
```suggestion
            RUNNER: self-hosted-1cpu
```

File: .github/workflows/smoke-test.yaml [hunk 17]
```diff
@@ -39,6 +39,17 @@ jobs:
             kv-gitlab-ci/data/github/shared/smoke-tests-registry-creds token_name ;
             kv-gitlab-ci/data/github/shared/smoke-tests-registry-creds token_secret ;
 
+      - name: Set cluster name
```
Comment (xDmitriev) [inline .github/workflows/smoke-test.yaml:42]: Remove it. It's unnecessary logic

File: .github/workflows/ci.yaml [hunk 18]
```diff
@@ -109,33 +106,38 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        k8s: [v1.23.13, v1.27.1]
-        ARCH: [amd64, arm64]
+        k8s: [ v1.23.13, v1.24.15, v1.25.11, v1.26.6, v1.27.1 ]
```
Comment (xDmitriev) [inline .github/workflows/ci.yaml:109]: we wanted to keep any 4 versions here.

File: .github/workflows/ci.yaml [hunk 19]
```diff
@@ -109,33 +106,38 @@ jobs:
...
             RUNNER: self-hosted-2cpu
-          - ARCH: arm64
-            RUNNER: github-runner-4cpu-arm64
     env:
       ARCH: ${{ matrix.ARCH }}
-      KIND_CLUSTER_NAME: kind-${{ matrix.k8s }}
       KUBECONFIG: $HOME/.kube/kind-config-${{ matrix.k8s }}
       SECRET: ${{ secrets.PGP_SECRET_SIGNING_PASSPHRASE }}
     steps:
+      - name: Set cluster name
```
Comment (xDmitriev) [inline .github/workflows/ci.yaml:119]: Same as below

File: .github/workflows/ci.yaml [hunk 20]
```diff
@@ -109,33 +106,38 @@ jobs:
...
         include:
           - ARCH: amd64
             RUNNER: self-hosted-2cpu
-          - ARCH: arm64
-            RUNNER: github-runner-4cpu-arm64
     env:
       ARCH: ${{ matrix.ARCH }}
-      KIND_CLUSTER_NAME: kind-${{ matrix.k8s }}
       KUBECONFIG: $HOME/.kube/kind-config-${{ matrix.k8s }}
       SECRET: ${{ secrets.PGP_SECRET_SIGNING_PASSPHRASE }}
```
Comment (xDmitriev) [inline .github/workflows/ci.yaml:117]: remove it

File: .github/workflows/ci.yaml [hunk 21]
```diff
@@ -109,33 +106,38 @@ jobs:
...
+        k8s: [ v1.23.13, v1.24.15, v1.25.11, v1.26.6, v1.27.1 ]
+        ARCH: [ amd64 ]
         include:
           - ARCH: amd64
             RUNNER: self-hosted-2cpu
-          - ARCH: arm64
-            RUNNER: github-runner-4cpu-arm64
     env:
       ARCH: ${{ matrix.ARCH }}
-      KIND_CLUSTER_NAME: kind-${{ matrix.k8s }}
```
Comment (xDmitriev) [inline .github/workflows/ci.yaml:121]: keep it

File: .github/workflows/ci.yaml [hunk 22]
```diff
@@ -109,33 +106,38 @@ jobs:
...
           name: controller-${{ env.ARCH }}.tar
-
       - name: Load images
         run: docker load -i controller-${{ env.ARCH }}.tar
-
       - name: Create cluster
-        run: kind create cluster --image=kindest/node:${{ matrix.k8s }}
-
+        run: |
+          echo "KIND_CLUSTER_NAME: $KIND_CLUSTER_NAME"
```
Comment (xDmitriev) [inline .github/workflows/ci.yaml:139]: same as before

File: test/smoke/run-smoke-suite.sh (hunk 12 above)
Comment (denisqsound) [reply test/smoke/run-smoke-suite.sh]: fixed

File: test/smoke/run.sh (hunk 14 above)
Comment (denisqsound) [reply test/smoke/run.sh]: changed to $(CURDIR)/allure_report

File: .github/workflows/smoke-test.yaml (hunk 15 above)
Comment (denisqsound) [reply .github/workflows/smoke-test.yaml]: changed

File: test/smoke/run-smoke-suite.sh (hunk 11 above)
Comment (denisqsound) [reply test/smoke/run-smoke-suite.sh]: removed

File: .github/workflows/ci.yaml (hunk 18 above)
Comment (denisqsound) [reply .github/workflows/ci.yaml]: removed v1.24.15 and v1.26.6

File: .github/workflows/ci.yaml (hunk 20 above)
Comment (denisqsound) [reply .github/workflows/ci.yaml]: removed

File: .github/workflows/ci.yaml (hunk 19 above)
Comment (denisqsound) [reply .github/workflows/ci.yaml]: removed

File: .github/workflows/smoke-test.yaml (hunk 17 above)
Comment (denisqsound) [reply .github/workflows/smoke-test.yaml]: removed

**Review by nedvna — APPROVED** [review]
Looking good

**Pull request merged**